### ISSUE TITLE
MPRIS Support

### ISSUE THREAD

**Issue by benjamindaines** [z-------------/CPod#28]
https://specifications.freedesktop.org/mpris-spec/latest/

Great podcast app!  Please consider adding MPRIS 2.0 support, which enables widget controls and general integration into Linux environments at some point.  Thanks.

**Comment by z-------------** [comment]
I'll look into it. Thanks 😊

**Comment by z-------------** [comment]
Hi, [(pre)release 1.20.0](https://github.com/z-------------/cumulonimbus/releases/tag/v1.20.0) contains basic MPRIS integration; a list of implemented MPRIS features can be found in the release notes. Tested on KDE neon. Please let me know if there are any problems.

**Comment by benjamindaines** [comment]
Seems to work.  Also glad to see that the feed is working now, I may be able to start testing this full time now :+1: 

Only thing is the metadata could use a little polishing for MPRIS.  I think it should display the podcast name where it currently displays "Unknown Artist" and then the episode title on the second line.  If you could also get the album cover (or whatever the podcast equivalent is) to show up there too it would be "on fleek" 

As always, awesome work! 
![screenshot from 2018-03-15 16-36-29](https://user-images.githubusercontent.com/3039460/37489884-71cc5e5a-286f-11e8-90a8-ce7a96566789.png)

**Comment by z-------------** [comment]
Got it. CPod currently puts the podcast title into MPRIS's album title field rather than the artist name field, so I'll switch it around.

As for the image, I'm guessing the reason it shows up empty is that the podcasts you have are in the old image storage format, which saves the image as a Blob directly into localForage, whereas the new format saves the images to disk as files. My understanding is that MPRIS's album image field only supports http(s):// and file:// URLs, and there is no clean and easy way to turn said blobs into either of those.

Assuming that this is the cause of the problem, I would recommend that you try Settings > Update podcast cover art. After that's done, the next played episode (or the same one, after a CPod restart) should now have the podcast image showing in the MPRIS controller.

I will investigate better solutions to this problem.

**Comment by z-------------** [comment]
[(Pre)release 1.20.1](https://github.com/z-------------/cumulonimbus/releases/tag/v1.20.1) includes the improvements you suggested. Please try it out and let me know if it fixes the problems for you.

**Comment by benjamindaines** [comment]
Something's still off there.  For some reason now it's back to unknown / unknown.  Album art didn't show up when it was displaying correct metadata thought.

**Comment by z-------------** [comment]
To clarify: the current situation is that no metadata or album art is showing?

**Comment by benjamindaines** [comment]
Neither album art nor metadata is show up.  Might it be gnome specific if it seems to be working for you in KDE?  I don't know why MPRIS would be different between the two DEs tho.

**Comment by z-------------** [comment]
Could you let me know what podcasts you're trying it with? Maybe there's a bug somewhere in my implementation.

**Comment by z-------------** [comment]
Here it is: [(pre)release 1.22.3](https://github.com/z-------------/cumulonimbus/releases/tag/v1.22.3)

**Comment by benjamindaines** [comment]
I'm not sure how to grab a link from cpod, but BBC's Planet Money is an example.

**Comment by z-------------** [comment]
I couldn't find any "Planet Money" from the BBC, I assume you mean the one from NPR? CPod's MPRIS integration seems to work fine with it, and the image update procedure I [mentioned earlier](https://github.com/z-------------/cumulonimbus/issues/28#issuecomment-373600158) seems to work fine as well. This might be hard to figure this out without those [logs](https://github.com/z-------------/cumulonimbus/issues/28#issuecomment-377521488).

One thing we could try, though, is to set the podcast title as MPRIS's artist name as well as the album name -- this latest problem seems to have arisen only after I switched those two fields around per your request. Maybe GNOME is seeing the missing album name and treating the metadata as invalid. Please try out [(pre)release 1.22.4](https://github.com/z-------------/cumulonimbus/releases/tag/v1.22.4) and let me know if that solves the problem.

**Comment by logix2** [comment]
The MPRIS implementation is a bit buggy for me. I'm using Ubuntu 18.04 and CPod creates MPRIS entries with the old application name in the [Media Player Indicator](https://extensions.gnome.org/extension/55/media-player-indicator/) extension each time it runs, here's a screenshot: https://i.imgur.com/VLA1bnW.png

**Comment by z-------------** [comment]
@benjamindaines It turns out that GNOME wants the MPRIS artist field as an array rather than a string, which is what CPod was giving it. 90e1bfb4a0bc30a07e02b32a017ec95f85205dd6 fixes this and this fix will be live in the next release of CPod:

![Screenshot](https://i.imgur.com/ClbPIui.png)

@logix2 f3173eb209e5c211197ba628db9ba98797433c94 fixes the old application name showing. I'll make a new issue for the duplicate entries bug.

![Screenshot](https://i.imgur.com/ylRLXdH.png)
